data_safe_haven/infrastructure/programs/sre/desired_state.py

```diff
@@ -0,0 +1,248 @@
+"""Pulumi component for SRE desired state"""
+
+from collections.abc import Mapping, Sequence
+
+import yaml
+from pulumi import (
+    ComponentResource,
+    FileAsset,
+    Input,
+    Output,
+    ResourceOptions,
+    StringAsset,
+)
+from pulumi_azure_native import (
+    network,
+    resources,
+    storage,
+)
+
+from data_safe_haven.external import AzureIPv4Range
+from data_safe_haven.functions import (
+    alphanumeric,
+    replace_separators,
+    sha256hash,
+    truncate_tokens,
+)
+from data_safe_haven.infrastructure.common import (
+    get_id_from_rg,
+    get_id_from_subnet,
+    get_name_from_rg,
+)
+from data_safe_haven.infrastructure.components import (
+    BlobContainerAcl,
+    BlobContainerAclProps,
+)
+from data_safe_haven.resources import resources_path
+from data_safe_haven.types import AzureDnsZoneNames
+
+
+class SREDesiredStateProps:
+    """Properties for SREDesiredStateComponent"""
+
+    def __init__(
+        self,
+        admin_ip_addresses: Input[Sequence[str]],
+        dns_private_zones: Input[dict[str, network.PrivateZone]],
+        gitea_hostname: Input[str],
+        hedgedoc_hostname: Input[str],
+        location: Input[str],
+        resource_group: Input[resources.ResourceGroup],
+        subscription_name: Input[str],
+        subnet_desired_state: Input[network.GetSubnetResult],
+    ) -> None:
+        self.admin_ip_addresses = admin_ip_addresses
+        self.dns_private_zones = dns_private_zones
+        self.gitea_hostname = gitea_hostname
+        self.hedgedoc_hostname = hedgedoc_hostname
+        self.location = location
+        self.resource_group_id = Output.from_input(resource_group).apply(get_id_from_rg)
+        self.resource_group_name = Output.from_input(resource_group).apply(
+            get_name_from_rg
+        )
+        self.subnet_desired_state_id = Output.from_input(subnet_desired_state).apply(
+            get_id_from_subnet
+        )
+        self.subscription_name = subscription_name
+
+
+class SREDesiredStateComponent(ComponentResource):
+    """Deploy SRE desired state with Pulumi"""
+
+    def __init__(
+        self,
+        name: str,
+        stack_name: str,
+        props: SREDesiredStateProps,
+        opts: ResourceOptions | None = None,
+        tags: Input[Mapping[str, Input[str]]] | None = None,
+    ) -> None:
+        super().__init__("dsh:sre:DesiredStateComponent", name, {}, opts)
+        child_opts = ResourceOptions.merge(opts, ResourceOptions(parent=self))
+        child_tags = {"component": "data"} | (tags if tags else {})
+
+        # Deploy desired state storage account
+        # - This holds the /desired_state container that is mounted by workspaces
+        # - Azure blobs have worse NFS support but can be accessed with Azure Storage Explorer
+        storage_account = storage.StorageAccount(
```

Comment:
Can we add a wrapped component (e.g. DSHStorageAccount) that would handle all the common storage account features so that we can easily see which options are varying between different accounts? I think we are creating at least 3, and it would be good to ensure that changes are kept in-sync.

Reply:
Maybe not for _all_ storage accounts.

This is one is a common archetype though, something like "NFSv3EnabledStorageAccount".
Maybe "V2StorageAccount" is another one.

Reply:
No, I mean we should create a new class that looks like:

```
class WrappedStorageAccount(StorageAccount):
  def __init__(
    self,
    a_parameter,
    a_second_parameter,
  ):
    super().__init__(
      common_parameter="value",
      common_parameter2="value2",
      a_parameter=a_parameter,
      a_second_parameter=a_second_parameter,
```

Reply:
Yes, but isn't it sensible to have a hierarchy. NFSv3EnabledStorageAccount and V2StorageAccount inherit from GenericStorageAccount.

Otherwise it is the same argument, we have multiple WrappedStorageAccounts which share common parameters.

Reply:
Sure, that's fine, but we'll only have one `NFSv3EnabledStorageAccount` and one `V2StorageAccount` right?

Reply:
There are two NFSv3 storage accounts. I think that is why I was more keen on abstracting that rather than creating a wrapper for all storage accounts (and as that is the kind of account this PR directly touches).